File: backend/app/models/sharepoint.py
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, HttpUrl

# Note: Aliases are used to map Pydantic fields to the camelCase keys
# returned by the Microsoft Graph API.

class SharePointSite(BaseModel):
    """Model representing a SharePoint site based on Graph API response."""
    id: str
    name: Optional[str] = None # Site name (sometimes not present in search results)
    display_name: str = Field(..., alias='displayName') # User-friendly display name
    web_url: str = Field(..., alias='webUrl') # URL to the site
    description: Optional[str] = None
    created_datetime: Optional[datetime] = Field(None, alias='createdDateTime')
    last_modified_datetime: Optional[datetime] = Field(None, alias='lastModifiedDateTime')

    class Config:
        populate_by_name = True # Allows using alias for population

class SharePointDrive(BaseModel):
    """Model representing a SharePoint document library (drive)."""
    id: str
    name: Optional[str] = None # Drive name
    drive_type: Optional[str] = Field(None, alias='driveType') # e.g., 'documentLibrary'
    web_url: str = Field(..., alias='webUrl') # URL to the drive
    description: Optional[str] = None
    created_datetime: Optional[datetime] = Field(None, alias='createdDateTime')
    last_modified_datetime: Optional[datetime] = Field(None, alias='lastModifiedDateTime')

    class Config:
        populate_by_name = True

class SharePointItem(BaseModel):
    """Model representing an item (file or folder) in a SharePoint drive."""
    id: str
    name: Optional[str] = None # File or folder name
    web_url: str = Field(..., alias='webUrl') # URL to the item
    size: Optional[int] = None # Size in bytes (present for files)
    created_datetime: Optional[datetime] = Field(None, alias='createdDateTime')
    last_modified_datetime: Optional[datetime] = Field(None, alias='lastModifiedDateTime')
    # These flags are added by the service layer, not from Graph directly
    is_folder: bool = Field(False, description="True if the item is a folder.") 
    is_file: bool = Field(False, description="True if the item is a file.")

    class Config:
        populate_by_name = True 

class SharePointDownloadRequest(BaseModel):
    """Request body for initiating a file download and processing task."""
    drive_id: str = Field(..., description="ID of the SharePoint drive containing the file.")
    item_id: str = Field(..., description="ID of the SharePoint file item to download.")
    # Optional: Add knowledge base ID or other processing parameters if needed
    # knowledge_base_id: Optional[str] = None 

# --- Models for Insights (Quick Access / Shared) ---

class ResourceReference(BaseModel):
    webUrl: Optional[HttpUrl] = None
    id: Optional[str] = None
    type: Optional[str] = None

class ResourceVisualization(BaseModel):
    title: Optional[str] = None
    type: Optional[str] = None # e.g., Word, Excel, PowerPoint, Pdf, etc.
    previewImageUrl: Optional[HttpUrl] = None
    # containerDisplayName: Optional[str] = None # Might be useful (e.g., Site name)
    # containerType: Optional[str] = None # e.g., "Site"

class LastUsedFacet(BaseModel):
    lastAccessedDateTime: Optional[datetime] = None
    lastModifiedDateTime: Optional[datetime] = None

class UsedInsight(BaseModel):
    id: str
    # lastUsed: Optional[LastUsedFacet] = None # Requires parsing lastUsed property which can be complex
    resourceVisualization: Optional[ResourceVisualization] = None
    resourceReference: Optional[ResourceReference] = None

class SharedInsight(BaseModel):
    id: str
    # sharingHistory: Optional[List[Any]] = None # Complex, skip for now
    resourceVisualization: Optional[ResourceVisualization] = None
    resourceReference: Optional[ResourceReference] = None 

File: frontend/src/models/sharepoint.ts
// Existing interfaces (SharePointSite, SharePointDrive, SharePointItem)
export interface SharePointSite {
    id: string;
    name?: string;
    displayName: string;
    webUrl: string;
}

export interface SharePointDrive {
    id: string;
    name?: string;
    driveType?: string;
    webUrl: string;
}

export interface SharePointItem {
    id: string;
    name?: string;
    webUrl: string;
    size?: number;
    createdDateTime?: string;
    lastModifiedDateTime?: string;
    is_folder: boolean;
    is_file: boolean;
}

// --- New Interfaces for Insights ---

export interface ResourceReference {
    webUrl?: string;
    id?: string;
    type?: string;
}

export interface ResourceVisualization {
    title?: string;
    type?: string; // e.g., Word, Excel etc.
    previewImageUrl?: string;
}

export interface UsedInsight {
    id: string;
    resourceVisualization?: ResourceVisualization;
    resourceReference?: ResourceReference;
    // Add lastUsed fields if needed later
}

// Placeholder for SharedInsight
export interface SharedInsight {
    id: string;
    resourceVisualization?: ResourceVisualization;
    resourceReference?: ResourceReference;
}

// --- Interfaces for Insights API (/insights/used) --- 

export interface InsightResourceReference {
    webUrl?: string;
    id?: string;
    type?: string;
}

export interface InsightResourceVisualization {
    title?: string;
    type?: string;
    previewImageUrl?: string;
}

export interface UsedInsight {
    id: string;
    resourceVisualization?: InsightResourceVisualization;
    resourceReference?: InsightResourceReference;
}

// +++ Interfaces for /me/drive/recent Endpoint +++

export interface Identity {
    displayName?: string;
    // id?: string;
}

export interface IdentitySet {
    user?: Identity;
    // application?: Identity;
    // device?: Identity;
}

export interface FileInfo {
    mimeType?: string;
}

export interface FolderInfo {
    childCount?: number;
}

// Renamed to avoid conflict with existing SharePointItem
export interface RecentDriveItem {
    id: string;
    name?: string;
    webUrl?: string;
    createdDateTime?: string; // Dates are strings from JSON
    lastModifiedDateTime?: string;
    size?: number;
    lastModifiedBy?: IdentitySet;
    file?: FileInfo;
    folder?: FolderInfo;
    // Helper properties can be added in component logic if needed
    // is_folder?: boolean;
    // last_modifier_name?: string;
}

// --- End Interfaces for /me/drive/recent ---

// --- Interfaces for SharePoint Sync List Feature (Phase 1) ---

export interface SharePointSyncItem {
    id: number; // DB primary key
    user_id: string; // Email of the user who added it
    item_type: 'file' | 'folder'; // Type of the item
    sharepoint_item_id: string; // ID from SharePoint
    sharepoint_drive_id: string; // Drive ID from SharePoint
    item_name: string; // Name of the item from SharePoint
}

// Used when adding an item (id and user_id are handled by backend)
export interface SharePointSyncItemCreate {
    item_type: 'file' | 'folder';
    sharepoint_item_id: string;
    sharepoint_drive_id: string;
    item_name: string;
}

// --- End Sync List Interfaces ---

// Placeholder for SharedInsight
export interface SharedInsight {
    id: string;
    resourceVisualization?: InsightResourceVisualization;
    resourceReference?: InsightResourceReference;
} 